src/core/parser/csv_parser/csv_parser.cpp

```diff
@@ -94,6 +103,16 @@ void CSVParser::GetLine(unsigned long long const line_index) {
 
 void CSVParser::GetNextIfHas() {
     has_next_ = !source_.eof();
+
+    /* Check for the last newline */
+    if (has_next_) {
+        source_.get();
+        if (source_.eof()) {
+            has_next_ = false;
+        }
+        source_.unget();
+    }
```

Comment:
A bit strange code given that right below there is an `if` with exactly the same condition, it's better to union them. Also, `get`, `eof`, `unget` sequence looks cumbersome and requires clearing of stream state in `PeekNext` which doesn't look right. Try something like this:
```cpp
    if (has_next_) {
        if (source_.peek() == std::ifstream::traits_type::eof()) {
            has_next_ = false;
            return;
        }
        GetNext();
    }
```